Extend NixGL compat

exzombie opened the pull request:
As promised in [my comment](https://github.com/nix-community/home-manager/pull/5355#issuecomment-2118819905) on nix-community/home-manager/pull/5355, here are the changes that make my use case work. I intend to dogfood this in the coming week.

I added a section to the user manual which explains why separate wrappers are needed. And, because there are two wrappers, I opted for removing default wrapping of firefox and kitty, at least for now. I believe this approach is simple enough for the user while being sufficiently flexible to cover both single-GPU and dual-GPU systems. Support for offloading to AMD dGPU is missing because I don't have such a system, but I see nothing that would prevent adding another wrapper if someone expresses the need and can test.

In the limited time I had, I focused on the functionality and user interface, but haven't checked conformance to the contributor guidelines. Sorry. I think we can fix that later. I'd like to get comments on the functionality first :)

Smona commented:
@exzombie I've force-pushed to remove the default wrappers, if you wouldn't mind rebasing your branch that would make it a bit easier to review :)

Smona commented on modules/misc/nixgl.nix:
```diff
@@ -39,26 +84,58 @@ in {
-            rm -rf $out/bin/*
             shopt -s nullglob # Prevent loop from running if no files
+
+            rm -rf $out/bin/*
```
good catch!

Smona commented on modules/misc/nixgl.nix:
```diff
@@ -39,26 +84,58 @@ in {
         }));
+  in {
+    lib.nixGL.wrapMesa = makePackageWrapper "Intel" "";
+    lib.nixGL.wrapNvidia = makePackageWrapper "Nvidia" nvOffloadEnv;
```
Shouldn't nvOffloadEnv inclusion here depend on the value of `cfg.nvidiaOffload`, like it does for the shell wrapper?

exzombie replied:
You're right, of course. A stupid mistake.

Smona commented on modules/misc/nixgl.nix:
```diff
@@ -4,30 +4,75 @@ let cfg = config.nixGL;
+    getWrapperExe = vendor: let
+      glPackage = cfg.packages.${pkgs.system}."nixGL${vendor}";
+      vulkanPackage = cfg.packages.${pkgs.system}."nixVulkan${vendor}";
+    in "${lib.getExe glPackage} ${lib.getExe vulkanPackage}";
```
why are you wrapping the process in both the glPackage and the vulkanPackage? shouldn't that be up to the user whether they want to use closed or open source drivers?

exzombie replied:
It's not about closed or open drivers. Vulkan and OpenGL are both graphics APIs, and it's up to the program which they use. I don't think the user should need to know which API the program uses, especially when the program can use either, as is the case of mpv or Chromium. See also https://github.com/nix-community/nixGL/issues/91

exzombie replied:
Honestly, I don't know why nixGL offers separate wrappers for OpenGL and Vulkan.

Smona commented on modules/misc/nixgl.nix:
```diff
@@ -39,26 +84,58 @@ in {
               chmod +x "$out/bin/$(basename $file)"
             done
+
+            # If .desktop files refer to the old package, replace the references
+            for dsk in "$out/share/applications"/*.desktop ; do
+              if ! grep "${pkg.out}" "$dsk" > /dev/null; then
+                continue
+              fi
+              src="$(readlink "$dsk")"
+              rm "$dsk"
+              sed "s|${pkg.out}|$out|g" "$src" > "$dsk"
+            done
```
awesome! 🥳

Smona reviewed (commented):
Thanks for the PR, and for adding docs!

Excuse me if I don't fully understand the intricacies of hybrid graphics laptop setup -- I tried playing around with it when I was a kid but it's been a long time. IIRC that was before vulkan.

I think this config schema might be a little difficult to use for someone who is sharing a module between multiple system configurations. One benefit of my original API that this doesn't have is that the `nixGL` package used to wrap all of your wrapped packages can be customized via a single per-system setting in flake.nix ([example](https://github.com/Smona/nixpkgs/blob/main/flake.nix#L96)).

What would you think of keeping that layer of indirection, but adding an option for a secondary GPU? the final settings would look something like:

- **packages**: like you have it now.
- **graphics**: one of "mesa", "nvidia", or "bumblebee". This would determine the nixGL wrapper `lib.nixGL.wrap` uses
- **offloadGraphics**: one of "mesa", "nvidia", or "bumblebee". This would be the same as `graphics`, but used for a secondary wrapper function, like `config.nixGL.wrapOffloaded`. Maybe `secondary` would be preferable to offload. This could default to the value of `graphics`. I think with "bumblebee" this would simply wrap the exe in `optirun`, but I'm not sure, so maybe bumblebee could be excluded for a first pass.
- **useVulkan**: bool, would switch nixGLNvidia->nixVulkanNvidia and nixGLIntel->nixVulkanIntel
- **nvidiaOffload**: I believe could stay the same as how you have it now, but I really don't know much about what these env vars do. Maybe we could do away with this option and just use the offloading env vars inside `wrapOffloaded` when the vendor is "nvidia"?
- **wrapperScripts**: could be a list of "mesa", "nvidia", and or "bumblebee" and work like it does now.

This seems like it would be a better compromise between our two use cases, and like it would avoid introducing any limitations that would require custom user settings & conditionals to work around. For example, someone could have a single configuration module for graphical apps that includes both firefox and blender. They `nixGL.wrap` firefox and `nixGL.wrapOffloaded` blender. This module could be well supported on a variety of systems just by tweaking options:

```nix
systems.nixos = myHmConfig {};
systems.ubuntuNotebook = myHmConfig {
  nixGL.packages = nixGL.packages;
  nixGL.graphics = "mesa";
};
systems.gamingLaptop = myHmConfig {
  nixGL.packages = nixGL.packages;
  nixGL.graphics = "mesa";
  nixGL.offloadGraphics = "nvidia";
};
systems.nvidiaDesktop = myHmConfig {
  nixGL.packages = nixGL.packages;
  nixGL.graphics = "nvidia";
  nixGL.useVulkan = true;
};
```

exzombie commented:
> This seems like it would be a better compromise between our two use cases, and like it would avoid introducing any limitations that would require custom user settings & conditionals to work around. For example, someone could have a single configuration module for graphical apps that includes both firefox and blender. They `nixGL.wrap` firefox and `nixGL.wrapOffloaded` blender.

I, too, maintain a single HM repo for multiple machines, both NixOS and non-NixOs. So, I'm well aware that there needs to be a way to configure things mostly the same but slightly differently. That's actually why I wanted to avoid defining a specific mechanism for this, as it seems to me that every user will have their own way of doing it. My intent was to provide wrappers, and allow the user to decide which wrapper to use in which situation.

That said, I can see the appeal of having a ready-made higher-level abstraction. I guess there's no harm in providing a primary+secondary wrapper mechanism, it would make it clear to users what the intended use is. But I'm pretty sure we should also give direct access to wrapper functions so that they can be used without indirection.

Speaking of wrapper functions, given the complexity of this stuff, there will have to be many. I thought we could provide just Mesa and Nvidia to start with and add others later as people ask for them, but I found that the offloading for Nvidia is complicated by the fact that there are different topologies, and what I provided will not work on all laptops. I'll see what I can come up with. I looked at the AMD offloading situation, and it seems it should be more straightforward.

exzombie commented:
Please tell me what you think of the new interface. I hope I managed to hit the sweet spot. For your case (single GPU, Mesa), it should work without any configuration apart from setting the packages option.

I looked into providing offloading on AMD and it's quite a bit simpler than for Nvidia, so I added it. I also made offloading a bit more flexible. Not that I can test, but at least the stuff is there so others can do so. I didn't want to include Bumblebee, though, because the way it's done in NixGL doesn't follow the same pattern as other wrappers. I don't know if it's worth it because I don't think the Bumblebee method is supported nowadays. If users appear, it can be added later.

Smona commented:
Awesome!! Thanks for being willing to work with me on the interface. I love where it's at now, really does feel like the best combination of convenience and flexibility.

I tried testing your branch with my configuration, and I got this error when building emacs:

```
Warning, nixVulkanIntel overwriting existing LD_LIBRARY_PATH
/nix/store/ynz405a4ngxijdmdz1qjmfb0ffgigx98-emacs-pgtk-29.1/bin/emacs: /nix/store/xpxln7rqi3pq4m0xpnawhxb2gs0mn1s0-gcc-12.3.0-lib/lib/libstdc++.so.6: version `GLIBCXX_3.4.31' not found (required by /nix/store/7byfnpsk04l3q8rvvwc06acwjbkiy0j2-webkitgtk-2.42.4+abi=4.0/lib/libwebkit2gtk-4.0.so.37)
/nix/store/ynz405a4ngxijdmdz1qjmfb0ffgigx98-emacs-pgtk-29.1/bin/emacs: /nix/store/xpxln7rqi3pq4m0xpnawhxb2gs0mn1s0-gcc-12.3.0-lib/lib/libstdc++.so.6: version `GLIBCXX_3.4.32' not found (required by /nix/store/7byfnpsk04l3q8rvvwc06acwjbkiy0j2-webkitgtk-2.42.4+abi=4.0/lib/libwebkit2gtk-4.0.so.37)
/nix/store/ynz405a4ngxijdmdz1qjmfb0ffgigx98-emacs-pgtk-29.1/bin/emacs: /nix/store/xpxln7rqi3pq4m0xpnawhxb2gs0mn1s0-gcc-12.3.0-lib/lib/libstdc++.so.6: version `GLIBCXX_3.4.31' not found (required by /nix/store/7byfnpsk04l3q8rvvwc06acwjbkiy0j2-webkitgtk-2.42.4+abi=4.0/lib/libjavascriptcoregtk-4.0.so.18)
```

It looks like the vulkan wrapper is overriding LD_LIBRARY_PATH with an incompatible version of GLIBC? I'll try and dig in to nixGL to get some more useful info, but I wonder if this is a sign that maybe wrapping with vulkan by default could make the wrappers work a little less consistently.

exzombie commented:
You're welcome! I'm glad nixGL is finally moving towards inclusion in HM.

Despite the fact that the warning uses the word "overwriting", I'm pretty sure it's just prepending, and not really throwing away the library path. It's also done by all wrappers, it's just the the Mesa Vulkan wrapper is the only one printing a warning.

If it's really the wrapper who pulls in an incompatible glibc (does the unwrapped package work?), it would mean that the libraries used by nixGL come from a different version of nixpkgs. Did you ensure that the nixGL input in your flake uses the same nixpkgs as HM? I can't reproduce this with 23.11

BTW, why are you wrapping emacs? Not that it matters for this problem, a wrapped package should always work, I'm just curious because I thought emacs didn't use a GPU.

Smona commented:
Hey @exzombie, I wish I hadn't taken so long to get back to you but life's been very busy. I'm still interested in merging your improvements before the nixgl module gets into home-manager, I just didn't want to add any blockers to the functionality getting merged. If we can get this ready before then, that would be great!

I wrapped emacs just out of an abundance of caution, trying to get as much performance out of it as possible. I'm using PGTK and I wasn't sure whether that could hardware accelerate rendering or not. Maybe it isn't actually helping, but I agree that the wrapper should avoid breaking packages as much as possible. I can confirm that my configuration builds and emacs runs on your branch when i unwrap emacs, all else being equal.

I believe everything is coming from the same version of nixpkgs, so I'm not sure what's causing the version mismatch. Here's my setup:

```nix
    # relevant flake inputs
    nixpkgs-ubuntu.url =
      "github:nixos/nixpkgs?rev=5ba549eafcf3e33405e5f66decd1a72356632b96";
 
    hm-ubuntu.url =
      "github:exzombie/home-manager-smona?rev=2b3674b62286dde05d0fcd816ee90b9f30003559";
    hm-ubuntu.inputs.nixpkgs.follows = "nixpkgs-ubuntu";

    nixGL.url = "github:guibou/nixGL";
    nixGL.inputs.nixpkgs.follows = "nixpkgs-ubuntu";

  # emacs.nix
  my-emacs = (config.lib.nixGL.wrap
    ((pkgs.emacsPackagesFor emacs).emacsWithPackages
      (epkgs: [ epkgs.vterm epkgs.pdf-tools ]))).overrideAttrs (oldAttrs: {
        # Temporarily working around this issue: https://github.com/NixOS/nixpkgs/issues/66706
        # TODO: submit upstream fix
        buildCommand = oldAttrs.buildCommand + ''
          ln -s $emacs/share/emacs $out/share/emacs
        '';
      });
```

I also tried updating nixGL to the latest version, which unsurprisingly didn't help. I have to pin a relatively old nixpkgs commit to maintain compatibility with Ubuntu's version of gnome, but I'm not sure where the incompatible GLIBC version could be sneaking in from. I will try to find some time to investigate further, but perhaps another option would be to omit the vulkan wrapper for now, or make it opt-in.

Let me know if I can get you any other info to try and reproduce my issue. My full configuration [is public](https://github.com/Smona/nixpkgs) if that helps.

This PR also needs to be rebased/merged with the updates to the wrapper script from other reviews on the base branch.

exzombie commented:
Today, I finally found the time to update my systems to 24.05. Now, I'm encountering the same problem. In my case, it manifests when a wrapped program tries to run executables from the host system, which fails because of missing glibc symbols. This is most unfortunate. I don't mind making Vulkan optional for now in order to not hold back this PR. However, I fear that OpenGL wrapping could cause similar issues when the stuff from nixpkgs deviates further from the stuff on the host system. In fact, I wonder why it doesn't cause trouble already ...

exzombie committed:
nixgl: API rework for flexibility and dual-GPU

exzombie commented:
I squashed when rebasing, and applied `./format`. Vulkan is now optional and disabled by default. I'm not entirely happy with the `makeWrapper` thing: it requires a single executable as wrapper, so, I added an intermediate wrapper. I don't think that's any nicer than what we had before. But let's see what the reviewers say.

AkechiShiro commented:
I take it that this PR would needs to be merged first ideally and then in turn we would be in a good place to merge upstream ? Maybe we should ask more people to test this branch, as I believe most are focused on the main branch that would land upstream

Thanks to you both, @exzombie and @Smona for the improvements of this very important feature for non-NixOS machine, this module may need to be activated by default on non-NixOS machine, at least I believe a working experience by default should be done instead of having a broken by default experience (which we currently have).

Smona commented on modules/misc/nixgl.nix:
```diff
@@ -1,29 +1,166 @@
+  config = let
+    getWrapperExe = vendor:
+      let
+        glPackage = cfg.packages.${pkgs.system}."nixGL${vendor}";
```
Will this work for non-flake setups? The manual claims the packages can be provided from a channel, but nixGL's `default.nix` doesn't seem to have this structure. 

I think this is my only remaining concern for the interface, that it couples this module to the internal structure of nixGL, which may change in the future. I wonder if we should keep the prefix-style arguments: `prefix`, (nullable) `offloadPrefix`, and (nullable) `vulkanPackage` instead of `vulkan.enable`.

We might need to drop the `installScripts` argument and `lib.wrappers` in that case, but I think that's totally fine, since users can already just pass their selected `nixGL` package into `home.packages` to get access to the wrapper script if they need it.

jackmac92 replied:
Is non flake support a requirement for all hm integrations? Or could that support be added after the initial merge?

jackmac92 replied:
And re: coupling to the internal structure of nixGL
the wrappers do seem useful, would adding an assertion in the wrapper about the expected structure of nixGL be a terrible idea?

exzombie replied:
You're right, of course; the structure of nixGL is convoluted. However, I see this HM module as a way to shield the user from this complexity, to a _resonable_ extent. When I first tried to integrate nixGL into my HM config, I spent a fair bit of time banging my head against the desk because stuff was documented as being under `auto`, but wasn't because I was using flakes. I'd like to spare people having to deal with that.

I think there is a way to deal with the difference in structure between channels and flakes in such a way that, should the attributes ever be moved from `auto` to the level above (which would make sense), the module won't break. However, I was unable to develop a solution because I hit a strange issue.

My own HM setup is a flake, so I wanted to spin up a VM to play with channels. So, I made a clean install of Debian (also tried Fedora), a clean setup of Nix using the "`curl | sh`" method, and found that nixGL doesn't build. For example, this command

```sh
NIXPKGS_ALLOW_UNFREE=1 nix --extra-experimental-features nix-command --extra-experimental-features flakes run --impure github:nix-community/nixGL#nixGLNvidia
```

yields the error 

```
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:34:12:
           33|
           34|   strict = derivationStrict drvAttrs;
             |            ^
           35|

       … while evaluating the derivation attribute 'name'
         at /nix/store/kc7nh9sdr4z8c7xc9j9891v212k2zj8k-source/pkgs/stdenv/generic/make-derivation.nix:270:7:
          269|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          270|       name =
             |       ^
          271|         let

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: cannot coerce null to a string: null
```

Similarly, trying to use channels by doing

```sh
nix-channel --add https://github.com/nix-community/nixGL/archive/main.tar.gz nixgl && nix-channel --update
nix-env -iA nixgl.auto.nixGLNvidia
```

results in

```
error:
       … in the condition of the assert statement
         at /nix/store/z7dky1g7nnxmfq7x6mz6xxswkyxmrflx-nixpkgs/nixpkgs/pkgs/build-support/trivial-builders/default.nix:89:5:
           88|     }:
           89|     assert lib.assertMsg (destination != "" -> (lib.hasPrefix "/" destination && destination != "/")) ''
             |     ^
           90|       destination must be an absolute path, relative to the derivation's out path,

       … in the left operand of the OR (||) operator
         at /nix/store/z7dky1g7nnxmfq7x6mz6xxswkyxmrflx-nixpkgs/nixpkgs/lib/asserts.nix:42:10:
           41|     msg:
           42|     pred || builtins.throw msg;
             |          ^
           43|

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: cannot coerce null to a string: null
```

I have no clue what's going on. It works fine on my system. Any ideas?

jackmac92 replied:
I've found I often need `--show-trace` to make sense of the error, otherwise the important parts get removed

I could try to recreate/debug, but want to follow the same steps for VM creation, do you have a link to a guide or something?

jackmac92 replied:
I think both y'all @exzombie and @Smona know far more about this than I do, but I'm here for (trying to) help. I used to use non-flake setup, can try to restore that, but slightly worried about the "re-efficacy" of my setup w/o some guidelines on a VM

exzombie replied:
I literally only installed a fresh Debian into a VM, did a multiuser install of nix, rebooted, then tried the above commands. They really should work on a clean install, methinks. I think I'm going crazy, so I would appreciate if you could at least reproduce the issue.

And yes, `--show-trace` is often useful, if verbose. Didn't help me any in this particular case.

jackmac92 replied:
I am working on trying to reproduce that issue, but running into some issues getting the virtual machine running, will update when I'm able to test it out

Smona replied:
As much as i hate that flakes are still considered experimental, and wish for them to be standardized soon, there are still some niche outstanding issues that might prevent someone from wanting to migrate to flakes yet ([lazy trees](https://github.com/NixOS/nix/issues/3121) come to mind), and ideally this module would support those users. However, I am a big fan of "shielding users from complexity" as @exzombie says, so I support the effort to provide a more ergonomic interface that supports both cases, if possible.

It looks like that second error is coming from `writeTextFile`, which is only used in `nixGL` for [writeExecutable](https://github.com/nix-community/nixGL/blob/310f8e49a149e4c9ea52f1adf70cdc768ec53f8a/nixGL.nix#L22). It suggests to me that a null name is somehow being passed there, but i don't see how that's possible. It might be coming from deeper in the stack.

My initial instinct is to write this off as "not worth root causing". There may be some limitations to your VM environment that prevent nixGL from running as expected. Could you instead test the non-flake version inside your flake config on bare metal, via `builtins.fetchGit`? I'm imagining something like this should work:

```nix
let nixGLSrc = builtins.fetchGit {
  url = "git@github.com:nix-community/nixGL.git";
  ref = "main";
  rev = "310f8e49a149e4c9ea52f1adf70cdc768ec53f8a";
};
nixGL = import "${nixGLSrc}/default.nx" { 
  pkgs = inputs.nixpkgs.legacyPackages.${system}; 
  # other configurations...
};
```

jackmac92 replied:
For testing the non-flake setup on bare-metal. I think having an existing flake install shouldn't matter, right? If that's not an issue I'll also test it out

exzombie replied:
> For testing the non-flake setup on bare-metal. I think having an existing flake install shouldn't matter, right? If that's not an issue I'll also test it out

I wish. I made a small home configuration, and got more weird evaluation errors. Searching shows similar errors occurred when people didn't pin their nixpgks properly. I must admit I do not have much interest in pursuing this further, seeing as I don't use channels.

@Smona, your "faux-channel" suggestion worked perfectly. I pushed a change which works both with a flake and with `default.nix`, and also extended the documentation with what I assume things would look like if one were to use an actual channel. This approach should keep working even if NixGL is ever refactored to move all packages to the same level.

exzombie committed:
nixgl: add support for channel-based configuration

Smona commented on modules/misc/nixgl.nix:
```diff
@@ -1,45 +1,185 @@
+        default), they are a no-op. This is useful on NixOS where the wrappers
+        are unnecessary.
+
+        Note that using any Nvidia wrapper requires building the configuration
+        with the `--impure` option.
```
TODO: link to the manual page here, trim down redundant info.

Smona commented on modules/misc/nixgl.nix:
```diff
@@ -1,45 +1,185 @@
+        shorthand.
+
+        The following wrappers are available:
+        ${wrapperListMarkdown}
```
nice :)

Smona reviewed (approved):
Very impressive & exciting work here @exzombie! I tested on my Ubuntu 24 intel system, setting only `nxGL.packages` and everything seems to be working just right.

I think this is ready to merge, just want to think it over for another day or so since there's a lot going on here. would you mind if I revise the docs a bit post-merge?

exzombie commented:
> I think this is ready to merge, just want to think it over for another day or so since there's a lot going on here. would you mind if I revise the docs a bit post-merge?

Go right ahead!

Smona commented on modules/misc/nixgl.nix:
```diff
@@ -53,11 +193,17 @@ in {
             ++ [ pkgs.makeWrapper ];
-          buildCommand = ''
+          buildCommand = let
+            # We need an intermediate wrapper package because makeWrapper
```
nitpick: I imagine that most people will leave Vulkan disabled, so we could probably avoid the double wrapper for most people with a little refactor. it would be a pretty small resource usage improvement though, probably not important

exzombie replied:
That is true, and I did consider it. But it seemed to me that the improvement would be too small compared to the effort needed for refactoring, especially because I was hoping at the time that the Vulkan situation could be improved and the default restored in the future. I have found in the meantime that it can't be improved, at least with the current approach used by NixGL that relies on environment variables. But this does not change the gain vs. effort equation.
